# Avivo Data Cleaning - Discharge Data

In [1]:
import pandas as pd
import numpy as np

### Loading the Discharge Data set:

In [2]:
discharge = pd.read_csv('C:/Users/prabl/Documents/Star82/Data/Discharge_Data_20180324T230158Z_001/discharge_data.csv')

In [3]:
print(discharge.shape)
discharge.head()

(1789, 67)


Form.Date                Program             Staff Client.Number  \
0  9/12/2014    Treatment - IDD Men      Qualley, Ann          4547   
1  9/26/2014    Treatment - IDD Men      Qualley, Ann     111155997   
2  9/29/2014  Treatment - IDD Women  Salsness, Carrie     111156277   
3  10/1/2014  Treatment - IDD Women  Salsness, Carrie     111156358   
4  10/1/2014  Treatment - IDD Women  Salsness, Carrie     111156017   

  Any.children.living.else.where.due.to.CPS.court.order.or.action  \
0  Not applicable - no children/no child protect ...                
1  Not applicable - no children/no child protect ...                
2  Not applicable - no children/no child protect ...                
3  Not applicable - no children/no child protect ...                
4  Not applicable - no children/no child protect ...                

  CD.Treatments.used  CHSR.Dimension.1  CHSR.Dimension.2  CHSR.Dimension.3  \
0                NaN        No problem        No problem  Moderate problem   
1                NaN        No problem        No problem     Minor problem   
2                NaN  Moderate problem     Minor problem  Moderate problem   
3                NaN        No problem   Serious problem  Moderate problem   
4                NaN     Minor problem  Moderate problem  Moderate problem   

   CHSR.Dimension.4            ...              \
0  Moderate problem            ...               
1  Moderate problem            ...               
2   Serious problem            ...               
3  Moderate problem            ...               
4   Extreme problem            ...               

     Tertiary.substance.abuse.problem Total.Charges..dollars.only.  \
0                             Unknown                          NaN   
1  No secondary or tertiary substance                         9270   
2  No secondary or tertiary substance                          NaN   
3            Other Opiates/Synthetics                          NaN   
4  No secondary or tertiary substance                          NaN   

             Victim.of.abuse Were.treatment.services.respectful.of.culture  \
0                         No                                       Unknown   
1                         No                                       Unknown   
2  Yes, physical and sexual                                        Unknown   
3                         No                                       Unknown   
4  Yes, physical and sexual                                        Unknown   

        Where.client.has.been.living.in.past.30.days  \
0  Independent living - including on own, self su...   
1  Independent living - including on own, self su...   
2  Independent living - including on own, self su...   
3  Dependent living - dependent children and/or a...   
4  Dependent living - dependent children and/or a...   

  Will.client.be.living.in.environment..conducive.to.recovery Date.of.Birth  \
0                                                Yes              2/10/1964   
1                                            Unknown              3/28/1985   
2                                            Unknown              7/20/1980   
3                                            Unknown              8/14/1991   
4                                            Unknown              6/13/1966   

    Age     Sex                       Race  
0  53.0    Male  Black or African American  
1  32.0    Male  Black or African American  
2  37.0  Female                      White  
3  26.0  Female  Black or African American  
4  51.0  Female  Black or African American  

[5 rows x 67 columns]

In [4]:
discharge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1789 entries, 0 to 1788
Data columns (total 67 columns):
Form.Date                                                                1789 non-null object
Program                                                                  1789 non-null object
Staff                                                                    1789 non-null object
Client.Number                                                            1789 non-null object
Any.children.living.else.where.due.to.CPS.court.order.or.action          1789 non-null object
CD.Treatments.used                                                       237 non-null object
CHSR.Dimension.1                                                         1789 non-null object
CHSR.Dimension.2                                                         1789 non-null object
CHSR.Dimension.3                                                         1789 non-null object
CHSR.Dimension.4                                

** We can see that there are a lot of columns(attributes) that have very few non-null values.  
Let's remove the columns that have more than 30% of data missing!**

Before computing the amount of missing data, let us replace all 'Unkown' with NA in the dataframe

In [5]:
discharge.replace('Unknown',np.nan,inplace=True)

In [6]:
Total_nulls = discharge.isnull().sum()
print(Total_nulls)

Form.Date                                                                   0
Program                                                                     0
Staff                                                                       0
Client.Number                                                               0
Any.children.living.else.where.due.to.CPS.court.order.or.action           454
CD.Treatments.used                                                       1552
CHSR.Dimension.1                                                            0
CHSR.Dimension.2                                                            0
CHSR.Dimension.3                                                            0
CHSR.Dimension.4                                                            0
CHSR.Dimension.5                                                            0
CHSR.Dimension.6                                                            0
CPS.involvement.results.from                                    

In [7]:
# computing percentage of missing data
total_rows = discharge.shape[0]
Total_nulls = (Total_nulls/total_rows)*100
print(Total_nulls)

Form.Date                                                                 0.000000
Program                                                                   0.000000
Staff                                                                     0.000000
Client.Number                                                             0.000000
Any.children.living.else.where.due.to.CPS.court.order.or.action          25.377306
CD.Treatments.used                                                       86.752376
CHSR.Dimension.1                                                          0.000000
CHSR.Dimension.2                                                          0.000000
CHSR.Dimension.3                                                          0.000000
CHSR.Dimension.4                                                          0.000000
CHSR.Dimension.5                                                          0.000000
CHSR.Dimension.6                                                          0.000000
CPS.

In [8]:
#Getting columns from dataframe that have less than 30% missing data
new_discharge = discharge.copy();

#Getting all columns that have LESS THAN 30% missing data
Validcols = Total_nulls <= 30 
print('Number of Columns that have sufficient data: ', Validcols.sum())

Number of Columns that have sufficient data:  31


In [9]:
#Getting the column names of these 28 attributes/columns
remaining_cols = Total_nulls[Validcols].index

#Checking if we have 28 column names
print(remaining_cols.shape)
remaining_cols

(31,)


Index(['Form.Date', 'Program', 'Staff', 'Client.Number',
       'Any.children.living.else.where.due.to.CPS.court.order.or.action',
       'CHSR.Dimension.1', 'CHSR.Dimension.2', 'CHSR.Dimension.3',
       'CHSR.Dimension.4', 'CHSR.Dimension.5', 'CHSR.Dimension.6',
       'CPS.involvement.results.from', 'Current.CD.Treatment',
       'Current.labor.force.status', 'Currently.involved.with.CPS',
       'Does.client.have.children', 'DSM.CD.Diagnosis',
       'DSM.CD.Diagnosis..Secondary.', 'Enrolled.in.school.or.job.training',
       'Number.of.children', 'Number.of.children.living.elsewhere',
       'Number.of.self.help.programs.attended.in.past.30.days',
       'Reason.for.Discharge', 'Secondary.substance.abuse.problem',
       'Substance.abuse.problem.at.Discharge',
       'Tertiary.substance.abuse.problem',
       'Where.client.has.been.living.in.past.30.days', 'Date.of.Birth', 'Age',
       'Sex', 'Race'],
      dtype='object')

In [10]:
#Generating the new dataframe
discharge_new = discharge[remaining_cols]
print(discharge_new.shape)
discharge_new.columns

(1789, 31)


Index(['Form.Date', 'Program', 'Staff', 'Client.Number',
       'Any.children.living.else.where.due.to.CPS.court.order.or.action',
       'CHSR.Dimension.1', 'CHSR.Dimension.2', 'CHSR.Dimension.3',
       'CHSR.Dimension.4', 'CHSR.Dimension.5', 'CHSR.Dimension.6',
       'CPS.involvement.results.from', 'Current.CD.Treatment',
       'Current.labor.force.status', 'Currently.involved.with.CPS',
       'Does.client.have.children', 'DSM.CD.Diagnosis',
       'DSM.CD.Diagnosis..Secondary.', 'Enrolled.in.school.or.job.training',
       'Number.of.children', 'Number.of.children.living.elsewhere',
       'Number.of.self.help.programs.attended.in.past.30.days',
       'Reason.for.Discharge', 'Secondary.substance.abuse.problem',
       'Substance.abuse.problem.at.Discharge',
       'Tertiary.substance.abuse.problem',
       'Where.client.has.been.living.in.past.30.days', 'Date.of.Birth', 'Age',
       'Sex', 'Race'],
      dtype='object')

In [11]:
discharge_new.sample(10)

Form.Date                  Program                 Staff Client.Number  \
863    5/20/2016      Treatment - IDD Men      Jackson, Jacinta          4116   
886     6/1/2016       Family Residential     Gross-Sand, Scott     111157239   
239     6/3/2015  Treatment - Relapse Men       Kamuelyu, David     111158024   
1244  12/30/2016      Treatment - IDD Men      Jackson, Jacinta     111159420   
968     7/8/2016  Treatment - Relapse Men      Jackson, Jacinta     111161233   
1479   6/23/2017      Treatment - IDD Men          Cox, Norman      111163949   
660     2/4/2016  Treatment - Relapse Men  Owusu-Asante, Thomas     111159493   
42     1/23/2015      Treatment - IDD Men          Qualley, Ann     111156034   
951     7/1/2016      Treatment - IDD Men  Owusu-Asante, Thomas     111161043   
131    3/17/2015      Treatment - IDD Men  Owusu-Asante, Thomas     111156715   

     Any.children.living.else.where.due.to.CPS.court.order.or.action  \
863   Not applicable - no children/no child protect ...                
886                                                  No                
239   Not applicable - no children/no child protect ...                
1244                                                NaN                
968   Not applicable - no children/no child protect ...                
1479                                                NaN                
660                                                 NaN                
42    Not applicable - no children/no child protect ...                
951   Not applicable - no children/no child protect ...                
131                                                 NaN                

      CHSR.Dimension.1  CHSR.Dimension.2  CHSR.Dimension.3  CHSR.Dimension.4  \
863         No problem        No problem     Minor problem        No problem   
886      Minor problem     Minor problem  Moderate problem   Serious problem   
239   Moderate problem     Minor problem  Moderate problem   Serious problem   
1244        No problem     Minor problem     Minor problem        No problem   
968         No problem        No problem  Moderate problem  Moderate problem   
1479        No problem     Minor problem  Moderate problem     Minor problem   
660   Unable to assess  Unable to assess  Unable to assess     Minor problem   
42          No problem        No problem  Moderate problem  Moderate problem   
951      Minor problem     Minor problem   Serious problem   Extreme problem   
131         No problem        No problem  Moderate problem   Serious problem   

      CHSR.Dimension.5                     ...                       \
863         No problem                     ...                        
886    Extreme problem                     ...                        
239    Serious problem                     ...                        
1244     Minor problem                     ...                        
968   Moderate problem                     ...                        
1479  Moderate problem                     ...                        
660      Minor problem                     ...                        
42    Moderate problem                     ...                        
951    Extreme problem                     ...                        
131    Extreme problem                     ...                        

     Number.of.self.help.programs.attended.in.past.30.days  \
863      16-30 times past month (0ver 3 times per week)      
886      1-3 times past month (less than once per week)      
239                                       No attendance      
1244     1-3 times past month (less than once per week)      
968      1-3 times past month (less than once per week)      
1479            Some attendance, but frequency unknown       
660      1-3 times past month (less than once per week)      
42                 4-7 times past month (once per week)      
951                                       No attendance      
131                                    

** We can see that we now have a dataset only with columns that have sufficient amount of data - in particular we are left with only 31 columns/attributes as compared to the original 67. **

Now, Looking at the columns/attributes to see if they make sence to keep in our dataset (that is they ar not dependant upon the admissions data)

In [12]:
list(discharge_new.columns)

['Form.Date',
 'Program',
 'Staff',
 'Client.Number',
 'Any.children.living.else.where.due.to.CPS.court.order.or.action',
 'CHSR.Dimension.1',
 'CHSR.Dimension.2',
 'CHSR.Dimension.3',
 'CHSR.Dimension.4',
 'CHSR.Dimension.5',
 'CHSR.Dimension.6',
 'CPS.involvement.results.from',
 'Current.CD.Treatment',
 'Current.labor.force.status',
 'Currently.involved.with.CPS',
 'Does.client.have.children',
 'DSM.CD.Diagnosis',
 'DSM.CD.Diagnosis..Secondary.',
 'Enrolled.in.school.or.job.training',
 'Number.of.children',
 'Number.of.children.living.elsewhere',
 'Number.of.self.help.programs.attended.in.past.30.days',
 'Reason.for.Discharge',
 'Secondary.substance.abuse.problem',
 'Substance.abuse.problem.at.Discharge',
 'Tertiary.substance.abuse.problem',
 'Where.client.has.been.living.in.past.30.days',
 'Date.of.Birth',
 'Age',
 'Sex',
 'Race']

Looks like there are multiple columns with CPS data - possibly redundant data?

In [13]:
#Taking a look at the CPS columns
discharge_new[['Currently.involved.with.CPS','CPS.involvement.results.from',
               'Any.children.living.else.where.due.to.CPS.court.order.or.action']]

Currently.involved.with.CPS  \
0     Not applicable - no children/no child protect ...   
1                                                    No   
2                                                    No   
3     Not applicable - no children/no child protect ...   
4     Not applicable - no children/no child protect ...   
5     Not applicable - no children/no child protect ...   
6                                                   NaN   
7                                                    No   
8     Not applicable - no children/no child protect ...   
9                                                   NaN   
10                                                  NaN   
11                                                   No   
12                                                   No   
13    Not applicable - no children/no child protect ...   
14                                                  NaN   
15                                                  NaN   
16    Not applicable - no children/no child protect ...   
17                                                  NaN   
18                                                   No   
19    Not applicable - no children/no child protect ...   
20                                                  NaN   
21                                                   No   
22                                                  NaN   
23    Not applicable - no children/no child protect ...   
24    Not applicable - no children/no child protect ...   
25                                                  NaN   
26    Not applicable - no children/no child protect ...   
27                                                   No   
28                                                  Yes   
29                                                   No   
...                                                 ...   
1759                                                 No   
1760                                                NaN   
1761  Not applicable - no children/no child protect ...   
1762                                                Yes   
1763                                                 No   
1764                                                Yes   
1765                                                 No   
1766  Not applicable - no children/no child protect ...   
1767  Not applicable - no children/no child protect ...   
1768  Not applicable - no children/no child protect ...   
1769                                                 No   
1770                                                Yes   
1771                                                Yes   
1772                                                NaN   
1773                                                 No   
1774                                                 No   
1775  Not applicable - no children/no child protect ...   
1776  Not applicable - no children/no child protect ...   
1777                                                 No   
1778                                                 No   
1779  Not applicable - no children/no child protect ...   
1780  Not applicable - no children/no child protect ...   
1781                                                NaN   
1782                                                Yes   
1783                                                NaN   
1784                                                NaN   
1785                                                 No   
1786                                                NaN   
1787                                                NaN   
1788                                                Yes   

                           CPS.involvement.results.from  \
0     Not applicable - no children/no child protect ...   
1     Not applicable - no children/no child protect ...   
2     Not applicable - no children/no child protect ...   
3     Not applicable - no children/no child protect ...   
4     Not applicable - no children/no child protect ...   
5     Not appli

Looking at these 3 columns we can see rows where data is inconsistent accross columns. 
For example: 
* Row 1771 - Yes on the first column but Unkown on the second column.
* Row 22 - Unkown on the first column but NA -no children/no CPS on the second column

** How do we consolidate the information accross the three columns into one column? **

Similarly, we have the columns: 'Does client have children' and 'Number of children' - These would essentially be conveying the same information.

In [14]:
#Looking at these two columns to confirm our hypothesis
discharge_new[['Does.client.have.children','Number.of.children']]

Does.client.have.children      Number.of.children
0                          Yes                     Two
1                          Yes                     One
2                          Yes                     One
3                           No  Client has no children
4                           No  Client has no children
5                          Yes                     Two
6                          Yes                   Three
7                          Yes                   Three
8                           No  Client has no children
9                          NaN                     NaN
10                          No  Client has no children
11                         Yes                   Three
12                         Yes                     One
13                         Yes                     Two
14                          No                     One
15                         NaN                     NaN
16                          No  Client has no children
17                         Yes                     Two
18                         Yes                     Two
19                          No                     One
20                          No                     Two
21                         Yes                     Two
22                          No  Client has no children
23                          No  Client has no children
24                          No                     NaN
25                         Yes                   Three
26                          No  Client has no children
27                         Yes                     Two
28                         Yes                     One
29                         Yes                     One
...                        ...                     ...
1759                       Yes                     Two
1760                       NaN                     NaN
1761                       Yes                    Four
1762                       Yes                    Four
1763                       Yes                     NaN
1764                       Yes                     One
1765                        No  Client has no children
1766                       Yes                     Two
1767                        No  Client has no children
1768                       Yes                   Three
1769                       Yes                     Two
1770                       Yes                    Four
1771                       Yes                     One
1772                       NaN                     NaN
1773                        No  Client has no children
1774                       Yes                    Four
1775                       Yes                     One
1776                        No  Client has no children
1777                       Yes                     Two
1778                       Yes                     Two
1779                        No  Client has no children
1780                        No  Client has no children
1781                       NaN                     NaN
1782                       Yes                   Three
1783                       NaN                     NaN
1784                       NaN                     NaN
1785                       Yes                   Three
1786                       NaN                     NaN
1787                       NaN                     NaN
1788                       Yes                     One

[1789 rows x 2 columns]

Looking at the data above we can see that the two columns are consistent with each other! We can now go ahead and **drop the column - 'Does client have children' ** as it does not add any extra information. 

Further, we can convert the categorical column **Number.of.children** into a numeric one - for later use.

In [15]:
#Dropping column
discharge_new = discharge_new.drop('Does.client.have.children', axis=1)
print(discharge_new.shape)

(1789, 30)


In [16]:
#Getting unique entries in the 'Number.of.children' column
discharge_new['Number.of.children'].unique()

array(['Two', 'One', 'Client has no children', 'Three', nan, 'Five',
       'Four', 'Ten or more', 'Six', 'Seven', 'Eight'], dtype=object)

In [17]:
#Converting 'Number.of.children' into numeric column
conv_dict = {'Number.of.children': {"One": 1, "Two" : 2, "Three":3, "Four":4,
                                   "Five":5, "Six":6,"Seven":7,"Eight":8, "Ten or more": 10,
                                   "Client has no children": 0, "Unknown": np.nan}}
discharge_new.replace(conv_dict, inplace=True)

In [18]:
print(discharge_new['Number.of.children'].dtypes)
discharge_new['Number.of.children'].unique()

float64


array([  2.,   1.,   0.,   3.,  nan,   5.,   4.,  10.,   6.,   7.,   8.])

In [19]:
print(discharge_new['Number.of.children.living.elsewhere'].dtype)

object


In [20]:
print(discharge_new['Number.of.children.living.elsewhere'].unique())

['No children/no child protect involvement' nan 'Ten or more' 'One' 'Four'
 'Two' 'Three' 'Six' 'Nine' 'Five' 'Seven']


In [21]:
#Converting these two columns also to numeric
conv_dict1 = {'Number.of.children.living.elsewhere': 
             {'No children/no child protect involvement':0, 'Ten or more':10, 'One':1,
              'Four':4, 'Two':2, 'Three':3, 'Six':6, 'Nine':9, 'Five':5, 'Seven': 7}}

discharge_new.replace(conv_dict1, inplace=True)

In [23]:
discharge_new[['Number.of.children','Number.of.children.living.elsewhere']]

Number.of.children  Number.of.children.living.elsewhere
0                    2.0                                  0.0
1                    1.0                                  0.0
2                    1.0                                  0.0
3                    0.0                                  0.0
4                    0.0                                  0.0
5                    2.0                                  0.0
6                    3.0                                  0.0
7                    3.0                                  0.0
8                    0.0                                  0.0
9                    NaN                                  NaN
10                   0.0                                  0.0
11                   3.0                                  0.0
12                   1.0                                  0.0
13                   2.0                                  0.0
14                   1.0                                  0.0
15                   NaN                                  NaN
16                   0.0                                  0.0
17                   2.0                                  NaN
18                   2.0                                  0.0
19                   1.0                                  0.0
20                   2.0                                  NaN
21                   2.0                                 10.0
22                   0.0                                 10.0
23                   0.0                                  0.0
24                   NaN                                  NaN
25                   3.0                                  NaN
26                   0.0                                  0.0
27                   2.0                                  0.0
28                   1.0                                  NaN
29                   1.0                                  0.0
...                  ...                                  ...
1759                 2.0                                  0.0
1760                 NaN                                  NaN
1761                 4.0                                  0.0
1762                 4.0                                  4.0
1763                 NaN                                  0.0
1764                 1.0                                  0.0
1765                 0.0                                  0.0
1766                 2.0                                  0.0
1767                 0.0                                  0.0
1768                 3.0                                  3.0
1769                 2.0                                  0.0
1770                 4.0                                  4.0
1771                 1.0                                  1.0
1772                 NaN                                  NaN
1773                 0.0                                  0.0
1774                 4.0                                  0.0
1775                 1.0                                  0.0
1776                 0.0                                  0.0
1777                 2.0                                  0.0
1778                 2.0                                  0.0
1779                 0.0                                  0.0
1780                 0.0                                  0.0
1781                 NaN                                  NaN
1782                 3.0                                  3.0
1783                 NaN                                  NaN
1784                 NaN                                  NaN
1785                 3.0                                  0.0
1786                 NaN                                  NaN
1787                 NaN                                  NaN
1788                 1.0                                  1.0

[1789 rows x 2 columns]

Looks like some rows have incorrect entries! If we look at row 22 and 21 the Number of children=0, but number of children living elsewhere is 10! Let's check if there are any more rows like this.

In [24]:
result = discharge_new[discharge_new['Number.of.children'] < discharge_new['Number.of.children.living.elsewhere']]
result['Number.of.children'].sum()

33.0

** Thus there are 33 rows where this conflict occurs! We do not know if this is an error or our understnading is wrong
Does Number.of.children.living.elsewhere include Non-biological kids as well?? **

In [25]:
# Calculating ratio of children lost to parental rights/living elsewhere might be 
# more useful as an indicator when compared to just the values
discharge_new['Ratio.living.elsewhere'] = -1.00
ratio = 0.0
for idx,row in discharge_new.iterrows():
    if row.loc['Number.of.children.living.elsewhere'] == 0 or row.loc['Number.of.children']==0:
        discharge_new.set_value(idx,'Ratio.living.elsewhere', 0)
        #print('here')
    elif pd.isnull(row.loc['Number.of.children.living.elsewhere']) or pd.isnull(row.loc['Number.of.children']):
        discharge_new.set_value(idx,'Ratio.living.elsewhere', -1)    
        #print('Here')
    else:
        ratio = row.loc['Number.of.children.living.elsewhere']/row.loc['Number.of.children']
        #print(ratio)
        discharge_new.set_value(idx,'Ratio.living.elsewhere', ratio)

discharge_new['Ratio.living.elsewhere'].replace(-1,np.nan,inplace=True)

In [26]:
discharge_new['Ratio.living.elsewhere'].unique()

array([ 0.        ,         nan,  5.        ,  0.5       ,  1.        ,
        0.75      ,  1.8       ,  3.33333333,  0.42857143,  0.66666667,
        0.57142857,  0.33333333,  0.25      ,  3.        ,  0.2       ,
        0.4       ,  2.        ,  0.8       ,  0.83333333,  0.625     ,
        0.85714286,  0.16666667,  0.6       ,  0.875     ])

In [27]:
#Taking a look at the new columns to make sure the computations are right
discharge_new[['Number.of.children','Number.of.children.living.elsewhere','Ratio.living.elsewhere']]

Number.of.children  Number.of.children.living.elsewhere  \
0                    2.0                                  0.0   
1                    1.0                                  0.0   
2                    1.0                                  0.0   
3                    0.0                                  0.0   
4                    0.0                                  0.0   
5                    2.0                                  0.0   
6                    3.0                                  0.0   
7                    3.0                                  0.0   
8                    0.0                                  0.0   
9                    NaN                                  NaN   
10                   0.0                                  0.0   
11                   3.0                                  0.0   
12                   1.0                                  0.0   
13                   2.0                                  0.0   
14                   1.0                                  0.0   
15                   NaN                                  NaN   
16                   0.0                                  0.0   
17                   2.0                                  NaN   
18                   2.0                                  0.0   
19                   1.0                                  0.0   
20                   2.0                                  NaN   
21                   2.0                                 10.0   
22                   0.0                                 10.0   
23                   0.0                                  0.0   
24                   NaN                                  NaN   
25                   3.0                                  NaN   
26                   0.0                                  0.0   
27                   2.0                                  0.0   
28                   1.0                                  NaN   
29                   1.0                                  0.0   
...                  ...                                  ...   
1759                 2.0                                  0.0   
1760                 NaN                                  NaN   
1761                 4.0                                  0.0   
1762                 4.0                                  4.0   
1763                 NaN                                  0.0   
1764                 1.0                                  0.0   
1765                 0.0                                  0.0   
1766                 2.0                                  0.0   
1767                 0.0                                  0.0   
1768                 3.0                                  3.0   
1769                 2.0                                  0.0   
1770                 4.0                                  4.0   
1771                 1.0                                  1.0   
1772                 NaN                                  NaN   
1773                 0.0                                  0.0   
1774                 4.0                                  0.0   
1775                 1.0                                  0.0   
1776                 0.0                                  0.0   
1777                 2.0                                  0.0   
1778                 2.0                                  0.0   
1779                 0.0                                  0.0   
1780                 0.0                                  0.0   
1781                 NaN                                  NaN   
1782                 3.0                                  3.0   
1783                 NaN                                  NaN   
1784                 NaN                                  NaN   
1785                 3.0                                  0.0   
1786                 NaN                                  NaN   
1787                 NaN                                  NaN   
1788                 1.0                 

In [28]:
# Checking percentage of null values in the ratio column
discharge_new['Ratio.living.elsewhere'].isnull().sum()/discharge_new.shape[0]*100

26.439351593068753

In [29]:
# Thus we can now drop the 'Number.of.children.living.elsewhere' column
discharge_new.drop('Number.of.children.living.elsewhere', axis=1,inplace=True)

Let us now look at the columns again:

In [30]:
discharge_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1789 entries, 0 to 1788
Data columns (total 30 columns):
Form.Date                                                          1789 non-null object
Program                                                            1789 non-null object
Staff                                                              1789 non-null object
Client.Number                                                      1789 non-null object
Any.children.living.else.where.due.to.CPS.court.order.or.action    1335 non-null object
CHSR.Dimension.1                                                   1789 non-null object
CHSR.Dimension.2                                                   1789 non-null object
CHSR.Dimension.3                                                   1789 non-null object
CHSR.Dimension.4                                                   1789 non-null object
CHSR.Dimension.5                                                   1789 non-null object
CHSR.Dimensio

## **Final points to consider: **

There are totally 30 columns remaining of the 67 original columns.

The following points were considered when removing columns:
* All columns with more than 30% missing data were removed - NaN and 'Unknown' were considered as missing data.
* Redundant column Does.client.have.children was removed and Number.of.children was retained. The column 'Number.of.children' was converted from categorical to Numeric.
* The column 'Number.of.children.living.elsewhere' column was dropped and the 'Ratio.living.elsewhere' was retained. However, there are a few rows that might have incorrect values as mentioned in the analysis above.
* The CPS columns are redundant. However, data is not consistent across columns and so I'm not sure of how to merge these columns